src/Xamarin.Android.Build.Tasks/Utilities/MarshalMethodsAssemblyRewriter.cs

```diff
@@ -0,0 +1,134 @@
+#if ENABLE_MARSHAL_METHODS
+using System;
+using System.Collections.Generic;
+using System.IO;
+
+using Java.Interop.Tools.Cecil;
+using Mono.Cecil;
+
+namespace Xamarin.Android.Tasks
+{
+	class MarshalMethodsAssemblyRewriter
+	{
+		IDictionary<string, MarshalMethodEntry> methods;
+		ICollection<AssemblyDefinition> uniqueAssemblies;
+		IDictionary <string, HashSet<string>> assemblyPaths;
+
+		public MarshalMethodsAssemblyRewriter (IDictionary<string, MarshalMethodEntry> methods, ICollection<AssemblyDefinition> uniqueAssemblies, IDictionary <string, HashSet<string>> assemblyPaths)
+		{
+			this.methods = methods ?? throw new ArgumentNullException (nameof (methods));
+			this.uniqueAssemblies = uniqueAssemblies ?? throw new ArgumentNullException (nameof (uniqueAssemblies));
+			this.assemblyPaths = assemblyPaths ?? throw new ArgumentNullException (nameof (assemblyPaths));
+		}
+
+		public void Rewrite (DirectoryAssemblyResolver resolver)
+		{
+			var unmanagedCallersOnlyAttributes = new Dictionary<AssemblyDefinition, CustomAttribute> ();
+			foreach (AssemblyDefinition asm in uniqueAssemblies) {
+				unmanagedCallersOnlyAttributes.Add (asm, GetUnmanagedCallersOnlyAttribute (asm, resolver));
+			}
+
+			Console.WriteLine ("Adding the [UnmanagedCallersOnly] attribute to native callback methods and removing unneeded fields+methods");
+			foreach (MarshalMethodEntry method in methods.Values) {
+				Console.WriteLine ($"\t{method.NativeCallback.FullName} (token: 0x{method.NativeCallback.MetadataToken.RID:x})");
+				method.NativeCallback.CustomAttributes.Add (unmanagedCallersOnlyAttributes [method.NativeCallback.Module.Assembly]);
+				method.Connector.DeclaringType.Methods.Remove (method.Connector);
+				method.CallbackField?.DeclaringType.Fields.Remove (method.CallbackField);
+			}
+
+			Console.WriteLine ();
```

Comment:
I think we'll be wanting `Log.LogDebugMessage()`, not `Console.WriteLine()`…

Reply:
The CWLs will be gone eventually, I will remove them in the last PR in the series